### DISCUSSION TITLE
Gihub Action builds Jupyter-Book PDF, how do I retrieve it?

### DISCUSSION THREAD

**Discussion by ratnanil** [executablebooks/meta#124] [Q&A] [answered]
I've successfully managed to initialize a github action to build the html version of my jupyter-book and push it to gh-pages. I've expaned the gh-action to also build the pdf version of my jupyter-book, which was also successful, since I received the message below. 
```
===============================================================================
A PDF of your book can be found at:
    /home/runner/work/codingingis/codingingis/_build/latex
===============================================================================
```

But I'm at a loss as how I can now **retrieve** this pdf. In my understanding, it is stored on the server instance which built my book, but it was not pushed back to my repo. How can I get the github action to do this?

**Comment by choldgraf** [comment, marked as answer]
perhaps @hamelsmu could provide guidance here?

**Reply by hamelsmu** [reply]
@ratnanil use the [upload artifact action](https://github.com/actions/upload-artifact)

**Reply by ratnanil** [reply]
That actually works quite well! But is there a way to push the pdf back into my repo? 

The issue with the upload artifact action is that the output is not publicly available via a predictable link. I.e. I cannot add a "[downlod the pdf version of this book](link_to_my_pdf)"-link to the book's website.

**Reply by ratnanil** [reply]
Found it! [github push action](https://github.com/ad-m/github-push-action)

Here's a snippet from my github action yaml (

```
    # Install the latex things to create the pdf
    - name: install latex dependencies
      run: |
        sudo apt-get install texlive-latex-recommended texlive-fonts-recommended texlive-latex-extra latexmk

    # Create the pdf version of the book
    - name: Build the pdf book
      run: |
        jupyter-book build . --builder pdflatex

    # This next step moves, renames and commits the created pdf
    - name: Commit and rename pdf
      run: |
        mv ./_build/latex/python.pdf codingingis.pdf
        git config --local user.email "action@github.com"
        git config --local user.name "GitHub Action"
        git add codingingis.pdf
        git commit -m "update PDF version of the book"
    - name: Push changes
      uses: ad-m/github-push-action@master
      with:
        github_token: ${{ secrets.GITHUB_TOKEN }}
```

**Reply by phockett** [reply]
Thanks @ratnanil, this is working nicely as a base layer for me, except I also had to add `texlive-xetex` to the "install dependencies" step for the default jupyter-book build configuration (as per https://jupyterbook.org/en/stable/advanced/pdf.html#build-a-pdf-using-latex).

For reference, my complex action can be found at https://github.com/phockett/Quantum-Metrology-with-Photoelectrons-Vol3/blob/main/.github/workflows/deploy.yml

**Reply by westurner** [reply]
From "How to build with the ReadTheDocs Docker container" https://github.com/ammaraskar/sphinx-action/issues/9 :
> The RTD containers have multiple versions of Python and e.g. all of LaTeX installed:
https://hub.docker.com/r/readthedocs/build/
